Code/Editor/EditorFramework/Assets/Implementation/AssetProcessor.cpp

```diff
@@ -384,114 +367,162 @@ bool ezProcessTask::GetNextAssetToProcess(ezUuid& out_guid, ezDataDirPath& out_p
 }
 
 
-void ezProcessTask::OnProcessCrashed()
+void ezProcessTask::OnProcessCrashed(ezStringView message)
 {
-  m_Status = ezStatus("Asset processor crashed");
+  ShutdownProcess();
+  m_Status = ezStatus(message);
   ezLogEntryDelegate logger([this](ezLogEntry& ref_entry) { m_LogEntries.PushBack(std::move(ref_entry)); });
-  ezLog::Error(&logger, "AssetProcessor crashed!");
-  ezLog::Error(&ezAssetProcessor::GetSingleton()->m_CuratorLog, "AssetProcessor crashed!");
+  ezLog::Error(&logger, message);
+  ezLog::Error(&ezAssetProcessor::GetSingleton()->m_CuratorLog, message);
 }
 
-bool ezProcessTask::BeginExecute()
+bool ezProcessTask::IsConnected()
 {
-  m_LogEntries.Clear();
-  m_TransitiveHull.Clear();
-  m_Status = ezStatus(EZ_SUCCESS);
-  {
-    EZ_LOCK(ezAssetCurator::GetSingleton()->m_CuratorMutex);
+  return m_pIPC->IsConnected();
+}
 
-    if (!GetNextAssetToProcess(m_AssetGuid, m_AssetPath))
-    {
-      m_AssetGuid = ezUuid();
-      m_AssetPath.Clear();
-      m_bDidWork = false;
-      return false;
-    }
+bool ezProcessTask::HasProcessCrashed()
+{
+  return m_pIPC->IsClientAlive();
+}
 
-    m_bDidWork = true;
-    ezAssetInfo::TransformState state = ezAssetCurator::GetSingleton()->IsAssetUpToDate(m_AssetGuid, nullptr, nullptr, m_uiAssetHash, m_uiThumbHash);
-    EZ_ASSERT_DEV(state == ezAssetInfo::TransformState::NeedsTransform || state == ezAssetInfo::TransformState::NeedsThumbnail, "An asset was selected that is already up to date.");
+bool ezProcessTask::Tick(bool bStartNewWork)
+{
+  while (true)
+  {
+    switch (m_State)
+    {
+      case State::LookingForWork:
+      {
+        if (!bStartNewWork)
+        {
+          return false; // don't call later
+        }
+        m_LogEntries.Clear();
+        m_TransitiveHull.Clear();
+        m_Status = ezStatus(EZ_SUCCESS);
+        {
+          EZ_LOCK(ezAssetCurator::GetSingleton()->m_CuratorMutex);
 
-    ezSet<ezString> dependencies;
+          if (!GetNextAssetToProcess(m_AssetGuid, m_AssetPath))
+          {
+            m_AssetGuid = ezUuid();
+            m_AssetPath.Clear();
+            return bStartNewWork; // call again if we should be looking for new work
+          }
 
-    ezStringBuilder sTemp;
-    ezAssetCurator::GetSingleton()->GenerateTransitiveHull(ezConversionUtils::ToString(m_AssetGuid, sTemp), dependencies, true, true);
+          ezAssetInfo::TransformState state = ezAssetCurator::GetSingleton()->IsAssetUpToDate(m_AssetGuid, nullptr, nullptr, m_uiAssetHash, m_uiThumbHash);
+          EZ_ASSERT_DEV(state == ezAssetInfo::TransformState::NeedsTransform || state == ezAssetInfo::TransformState::NeedsThumbnail, "An asset was selected that is already up to date.");
```

Comment:
not sure whether I would assert on those states, there are so many things that can interfere and produce race conditions ...

Reply:
This is not new code, that was always there. Also all ezProcessTask instaces run in the same thread sequentially, so at least the ezProcessTasks can't interfere with each other.